From 87b40fc13f4b65631f9f6ed622ffe13fe31025b1 Mon Sep 17 00:00:00 2001
From: Ghennadi Procopciuc <ghennadi.procopciuc@nxp.com>
Date: Wed, 31 Aug 2022 10:41:29 +0300
Subject: [PATCH 084/102] arm: s32: Remove cluster1 when running in lokstep

Remove cluster1 node from cpu-map node when booting the
OS in lockstep mode.

Issue: ALB-9246
Upstream-Status: Pending 

Signed-off-by: Ghennadi Procopciuc <ghennadi.procopciuc@nxp.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 arch/arm/mach-s32/s32-cc/fdt.c | 18 ++++++++++++++----
 1 file changed, 14 insertions(+), 4 deletions(-)

diff --git a/arch/arm/mach-s32/s32-cc/fdt.c b/arch/arm/mach-s32/s32-cc/fdt.c
index a9f744f5d0..5a3d6881e4 100644
--- a/arch/arm/mach-s32/s32-cc/fdt.c
+++ b/arch/arm/mach-s32/s32-cc/fdt.c
@@ -21,7 +21,8 @@
 #include <dt-bindings/nvmem/s32cc-siul2-nvmem.h>
 #include <dt-bindings/phy/phy.h>
 
-#define S32_DDR_LIMIT_VAR "ddr_limitX"
+#define S32_DDR_LIMIT_VAR	"ddr_limitX"
+#define FDT_CLUSTER1_PATH	"/cpus/cpu-map/cluster1"
 
 static int get_core_id(u32 core_mpidr, u32 max_cores_per_cluster)
 {
@@ -120,12 +121,12 @@ static bool is_lockstep_enabled(void)
 
 static int ft_fixup_cpu(void *blob)
 {
-	int ret, off, addr_cells = 0;
+	int ret, addr_cells = 0;
 	u32 max_cores_per_cluster = 0;
 	u32 cpu_mask = 0;
 	u64 core_mpidr, core_id;
 	fdt32_t *reg;
-	int off_prev;
+	int off, off_prev, cluster1_off;
 
 	ret = get_cores_info(&max_cores_per_cluster, &cpu_mask);
 	if (ret)
@@ -167,7 +168,16 @@ static int ft_fixup_cpu(void *blob)
 		off = get_next_cpu(blob, off);
 	}
 
-	return 0;
+	if (!is_lockstep_enabled())
+		return 0;
+
+	cluster1_off = fdt_path_offset(blob, FDT_CLUSTER1_PATH);
+	if (cluster1_off < 0) {
+		printf("couldn't find %s node\n", FDT_CLUSTER1_PATH);
+		return -ENODEV;
+	}
+
+	return fdt_del_node(blob, cluster1_off);
 }
 
 static int apply_memory_fixups(void *blob, bd_t *bd)
-- 
2.17.1

